aesd-char-driver/main.c

```diff
@@ -205,13 +210,167 @@ ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
+    }
+    mutex_unlock(&dev->lock);
+    return retval;
+}
+/*
+ * @function	:  ioctl description for AESDCHAR_IOCSEEKTO command
+ *
+ * @param		: cmd :Command to be passed to ioctl defined in aesd_ioctl.h , arg :any arguments passed with the command
+ * @return		:  retval :error condition
+ *
+ */
+long aesd_ioctl(struct file *filp, unsigned int cmd, unsigned long arg)
+{
+
+    int err = 0;
+    long retval = 0;
+    struct aesd_seekto seekto;
+    if (filp == NULL)
+    {
+
+        return -EFAULT;
+    }
+    PDEBUG("IOCTL sys call\n");
+    /*
+     * extract the type and number bitfields, and don't decode
+     * wrong cmds: return ENOTTY (inappropriate ioctl) before access_ok()
+     */
+    if (_IOC_TYPE(cmd) != AESD_IOC_MAGIC)
+        return -ENOTTY;
+    if (_IOC_NR(cmd) > AESDCHAR_IOC_MAXNR)
+        return -ENOTTY;
+
+    /*
+     * access_ok is kernel-oriented,for checking "read" and "write" access
+     */
+    if (_IOC_DIR(cmd) & _IOC_READ)
+        err = !access_ok((void __user *)arg, _IOC_SIZE(cmd));
+    else if (_IOC_DIR(cmd) & _IOC_WRITE)
+        err = !access_ok((void __user *)arg, _IOC_SIZE(cmd));
+    if (err)
+        return -EFAULT;
+
+    switch (cmd)
+    {
+
+    case AESDCHAR_IOCSEEKTO:
+       
+        if (copy_from_user(&seekto, (const void __user *)arg, sizeof(seekto))) // check if command from userspace copied to kernel space
+           { PDEBUG("Error while copying from userspace\n");
+            retval = -EFAULT;  }                                                // if it returns non zero means error
+        else
+        {
+            PDEBUG("Implementing AESDCHAR_IOCSEEKTO\n");
+            retval = aesd_adjust_file_offset(filp, seekto.write_cmd, seekto.write_cmd_offset);
+        }
+        break;
+
+    default: 
+        return -ENOTTY;
+    }
+    return retval;
+}
+/*
+ * @function	:  implementation of llseek in kernel space
+ *
+ * @param		: offset: offset to check for , whence : int to indicate which type of seek (SEEK_SET, SEEK_CUR, and SEEK_END)
+ * @return		:  loff_t seek_pos: file position after seek
+ *
+ */
+loff_t aesd_llseek(struct file *filp, loff_t offset, int whence)
+{
+
+    struct aesd_dev *dev = NULL;
+    struct aesd_buffer_entry *seek_entry = NULL;
+    loff_t buffer_size = 0;
+    loff_t seek_pos = 0;
+    uint8_t index = 0;
+
+    PDEBUG("llseek implementation\n");
+
+    // checking for error condition
+    if (filp == NULL)
+    {
+
+        return -EFAULT;
+    }
+
+    dev = (struct aesd_dev *)filp->private_data;
+
+    // lock the mutex
+    if (mutex_lock_interruptible(&(dev->lock)))
+    {
+        PDEBUG(KERN_ERR "could not acquire mutex lock");
+        return -ERESTARTSYS;
+    }
+
+    // getting size of buffer
+    AESD_CIRCULAR_BUFFER_FOREACH(seek_entry, &dev->circle_buff, index)
+    {
+        buffer_size += seek_entry->size;
+    }
+        //have used the fixed_size_llseek() function 
+    seek_pos = fixed_size_llseek(filp, offset, whence, buffer_size);
```

Comment:
@ayswariya088 do you need to hold the dev lock for the call to `fixed_size_llseek`?